File: backend/app/schemas/progress.py
from datetime import datetime

from pydantic import BaseModel


class LessonProgressOut(BaseModel):
    lesson_id: int
    completed: bool
    completed_at: datetime | None


class CourseProgressOut(BaseModel):
    course_id: int
    course_title: str
    total_lessons: int
    completed_lessons: int
    percentage: float
    certificate_number: str | None = None


class OverallProgressOut(BaseModel):
    courses: list[CourseProgressOut]
    overall_percentage: float

File: mobile/lib/models/progress.dart
class CourseProgress {
  final int courseId;
  final String courseTitle;
  final int totalLessons;
  final int completedLessons;
  final double percentage;

  CourseProgress({
    required this.courseId,
    required this.courseTitle,
    required this.totalLessons,
    required this.completedLessons,
    required this.percentage,
  });

  factory CourseProgress.fromJson(Map<String, dynamic> json) => CourseProgress(
    courseId: json['course_id'] as int,
    courseTitle: json['course_title'] as String,
    totalLessons: json['total_lessons'] as int,
    completedLessons: json['completed_lessons'] as int,
    percentage: (json['percentage'] as num).toDouble(),
  );
}

class OverallProgress {
  final List<CourseProgress> courses;
  final double overallPercentage;

  OverallProgress({required this.courses, required this.overallPercentage});

  factory OverallProgress.fromJson(Map<String, dynamic> json) =>
      OverallProgress(
        courses: (json['courses'] as List)
            .map((e) => CourseProgress.fromJson(e as Map<String, dynamic>))
            .toList(),
        overallPercentage: (json['overall_percentage'] as num).toDouble(),
      );
}
